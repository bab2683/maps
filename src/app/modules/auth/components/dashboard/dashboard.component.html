<ng-container *ngIf="user$ | async as user; else showLogin">
  <auth-logged-in [user]="user" (logout)="logout()"></auth-logged-in>
</ng-container>
<ng-template #showLogin>
  <auth-login (loginData)="login($event)"></auth-login>
</ng-template>
