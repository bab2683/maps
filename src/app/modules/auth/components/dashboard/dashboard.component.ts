import { Component } from '@angular/core';
import { Observable } from 'rxjs';

import { AuthService } from '../../auth.service';

@Component({
  selector: 'auth-dashboard',
  templateUrl: './dashboard.component.html',
  styleUrls: ['./dashboard.component.scss']
})
export class DashboardComponent {
  isLoggedin: Observable<boolean>;

  constructor(private auth: AuthService) {
    this.isLoggedin = this.auth.isLoggedIn();
  }
}
