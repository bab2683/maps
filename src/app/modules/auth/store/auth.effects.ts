import { Injectable } from '@angular/core';
import { AngularFireAuth } from '@angular/fire/auth';
import { Actions, Effect, ofType } from '@ngrx/effects';
import { FirebaseError } from 'firebase';
import { from, Observable, of } from 'rxjs';
import { catchError, first, switchMap } from 'rxjs/operators';

import { getRelevantUserData, parseLoginError } from '../utils';
import {
  AuthActionTypes,
  AuthInitCheck,
  AuthLoadedAction,
  AuthLoginFail,
  AuthLoginRequest,
  AuthLoginSuccess,
  AuthLogoutRequest,
  AuthLogoutSuccess
} from './auth.actions';

@Injectable()
export class AuthEffects {
  @Effect()
  public initAuth$: Observable<AuthLoadedAction> = this.actions$.pipe(
    ofType<AuthInitCheck>(AuthActionTypes.INIT_CHECK),
    switchMap(() => {
      return this.afAuth.user.pipe(
        first(),
        switchMap(user => {
          return of(new AuthLoadedAction(user ? getRelevantUserData(user) : null));
        })
      );
    })
  );

  @Effect()
  public loginRequest$: Observable<AuthLoginFail | AuthLoginSuccess> = this.actions$.pipe(
    ofType<AuthLoginRequest>(AuthActionTypes.LOGIN_REQUEST),
    switchMap(({ payload: { email, password } }) =>
      from(this.afAuth.auth.signInWithEmailAndPassword(email, password)).pipe(
        switchMap(({ user }) => {
          return of(new AuthLoginSuccess(getRelevantUserData(user)));
        }),
        catchError((err: FirebaseError) => {
          return of(new AuthLoginFail(parseLoginError(err)));
        })
      )
    )
  );

  @Effect()
  public logoutRequest$: Observable<AuthLogoutSuccess> = this.actions$.pipe(
    ofType<AuthLogoutRequest>(AuthActionTypes.LOGOUT_REQUEST),
    switchMap(() =>
      from(this.afAuth.auth.signOut()).pipe(switchMap(() => of(new AuthLogoutSuccess())))
    )
  );

  constructor(private actions$: Actions, private afAuth: AngularFireAuth) {}
}
