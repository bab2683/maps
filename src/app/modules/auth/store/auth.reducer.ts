import { AuthActions, AuthActionTypes } from './auth.actions';
import { AuthState, initialAuthState } from './auth.state';

export function authReducer(state: AuthState = initialAuthState, action: AuthActions): AuthState {
  switch (action.type) {
    case AuthActionTypes.LOADED: {
      return {
        ...state,
        loading: false,
        user: action.payload
      };
    }

    case AuthActionTypes.LOGIN_SUCCESS: {
      return {
        ...state,
        user: action.payload
      };
    }

    case AuthActionTypes.LOGIN_ERROR: {
      return {
        ...state,
        error: action.payload
      };
    }

    case AuthActionTypes.LOGOUT_SUCCESS: {
      return {
        ...state,
        user: null
      };
    }

    default:
      return state;
  }
}
