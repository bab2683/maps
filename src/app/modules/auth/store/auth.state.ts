import { AuthError, LoggedUser } from '../auth.model';

export interface AuthState {
  error?: AuthError;
  loading: boolean;
  user: LoggedUser | null;
}

export const initialAuthState: AuthState = {
  loading: true,
  user: null
};

export const AUTH_STATE = 'auth';
