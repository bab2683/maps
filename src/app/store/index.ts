export * from './app.actions';
export * from './app.effects';
export * from './app.reducer';
export * from './app.state';
export * from './ui';
