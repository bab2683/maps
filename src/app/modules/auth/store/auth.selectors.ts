import { createFeatureSelector, createSelector } from '@ngrx/store';

import { AUTH_STATE, AuthState } from './auth.state';

export const getAuthState = createFeatureSelector<AuthState>(AUTH_STATE);

export const getAuthLoadingState = createSelector(
  getAuthState,
  ({ loading }: AuthState) => loading
);

export const getLoggedUser = createSelector(
  getAuthState,
  ({ user }: AuthState) => user
);
