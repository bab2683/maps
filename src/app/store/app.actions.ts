import { Action } from '@ngrx/store';

export enum AppActionTypes {
  LOADED = '[App] loaded'
}

export class AppLoadedAction implements Action {
  public readonly type: string = AppActionTypes.LOADED;
}

export type AppActions = AppLoadedAction;
