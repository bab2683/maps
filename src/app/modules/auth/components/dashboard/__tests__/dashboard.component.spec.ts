import { NO_ERRORS_SCHEMA } from '@angular/compiler/src/core';
import { ComponentFixture, TestBed } from '@angular/core/testing';
import { Store } from '@ngrx/store';
import { MockStore, provideMockStore } from '@ngrx/store/testing';

import {
  AuthState,
  AuthInitCheck,
  AuthLoginRequest,
  AuthLogoutRequest,
  getLoggedUser,
  initialAuthState
} from '../../../store';

import { DashboardComponent } from '../dashboard.component';

describe('DashboardComponent', () => {
  let component: DashboardComponent;
  let fixture: ComponentFixture<DashboardComponent>;
  let store: MockStore<AuthState>;
  let dispatchSpy: any;
  let selectSpy: any;

  beforeEach(() => {
    TestBed.configureTestingModule({
      declarations: [DashboardComponent],
      providers: [provideMockStore<AuthState>({ initialState: initialAuthState })],
      schemas: [NO_ERRORS_SCHEMA]
    }).compileComponents();
    store = TestBed.get<Store<AuthState>>(Store);
    dispatchSpy = jest.spyOn(store, 'dispatch');
    selectSpy = jest.spyOn(store, 'select');
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(DashboardComponent);
    component = fixture.componentInstance;
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should match snapshot', () => {
    expect(fixture).toMatchSnapshot();
  });

  describe('onInit', () => {
    beforeEach(() => {
      component.ngOnInit();
    });

    it('should dispatch init auth action', () => {
      const action = new AuthInitCheck();
      expect(dispatchSpy).toBeCalledWith(action);
    });

    it('should fetch user data', () => {
      expect(selectSpy).toBeCalledWith(getLoggedUser);
    });
  });

  describe('login', () => {
    it('should dispatch login request action', () => {
      const credentials = { email: '', password: '' };
      component.login(credentials);
      const action = new AuthLoginRequest(credentials);
      expect(dispatchSpy).toBeCalledWith(action);
    });
  });

  describe('logout', () => {
    it('should dispatch logout request action', () => {
      component.logout();
      const action = new AuthLogoutRequest();
      expect(dispatchSpy).toBeCalledWith(action);
    });
  });
});
