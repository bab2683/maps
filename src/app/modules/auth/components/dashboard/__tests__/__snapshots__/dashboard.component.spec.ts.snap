// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DashboardComponent should match snapshot 1`] = `
<auth-dashboard
  store={[Function MockStore]}
>
  
</auth-dashboard>
`;
